# 파이썬 기본 자료형 2부

## 주요 내용 

* 스칼라 자료형 2편: 문자열(`str`)

## 스칼라 자료형 2편

스칼라 자료형 1편에서 소개한 정수, 부동소수점, 부울 값, `None` 값과 함께 스칼라 자료형에 속하는
문자열 자료형을 자세히 살펴본다.

## 문자열 자료형: `str`

문자열은 작은따옴표(`'`) 또는 큰따옴표(`"`)를 사용한다. 

In [1]:
a = '작은따옴표를 사용하는 문자열'

a

'작은따옴표를 사용하는 문자열'

In [2]:
b = "큰따옴표를 사용하는 문자열"

b

'큰따옴표를 사용하는 문자열'

여러 줄로 이루어진 문자열은 삼중 큰따옴표로 감싼다.

In [3]:
c = """여러 줄에 걸친 문자열은
    삼중 큰따옴표로 감싼다."""

c

'여러 줄에 걸친 문자열은\n    삼중 큰따옴표로 감싼다.'

### 인덱스, 인덱싱, 슬라이싱

`text` 변수가 다음 긴 문자열을 가리킨다.

In [4]:
text = """<html><head><title>Welcome to the Beans'R'Us Pricing Page</title>
<link rel="stylesheet" type="text/css" href="beansrus.css" />
</head><body>
<h2>Welcome to the Beans'R'Us Pricing Page</h2>
<p>Current price of coffee beans = <strong>$5.01</strong></p>
<p>Price valid for 15 minutes from Sat Sep 9 09:43:38 2017.</p>
</body></html>
"""

In [5]:
print(text)

<html><head><title>Welcome to the Beans'R'Us Pricing Page</title>
<link rel="stylesheet" type="text/css" href="beansrus.css" />
</head><body>
<h2>Welcome to the Beans'R'Us Pricing Page</h2>
<p>Current price of coffee beans = <strong>$5.01</strong></p>
<p>Price valid for 15 minutes from Sat Sep 9 09:43:38 2017.</p>
</body></html>



__참고:__ 위 텍스트는 [Coffee Beans Pricing Page](http://beans.itcarlow.ie/prices.html)
웹페이지의 소스코드이다. (크롬 브라우저에서 F12키를 누르면 확인할 수 있다.)

<div align="center"><img src="images/coffee-beans04.jpg" width="700"/></div>

__질문:__ 위 문자열에서 `$5.01` 정보를 어떻게 끄집어 낼 수 있을까?

__정답:__ 인덱스, 인덱싱, 슬라이싱을 활용한다.

### 인덱스

문자열에 사용되는 모든 문자의 위치는 __인덱스__(index)라는 고유한 번호를 갖는다.
인덱스는 0부터 시작하며 오른쪽으로 한 문자씩 이동할 때마다 1씩 증가한다. 

**주의:** 파이썬을 포함해서 많은 대부분의 프로그래밍 언어에서 인덱스는 0부터 시작한다. 
따라서 첫 째 문자를 확인하고자 할 때는 1이 아닌 0을 인덱스로 사용해야 한다. 

예제를 통해 인덱스에 친숙해질 필요가 있다.

In [6]:
a_food = "kebap"

### 인덱싱

특정 인덱스에 위치한 문자는 다음과 같이 확인한다.

* 0번 인덱스 값, 즉 첫째 문자

In [7]:
a_food[0]

'k'

* 1번 인덱스를 갖는 값, 즉, 둘째 문자

In [8]:
a_food[1]

'e'

* 2번 인덱스를 갖는 값, 즉, 셋째 문자

In [9]:
a_food[2]

'b'

등등.

#### -1번 인덱스

문자열이 많이 긴 경우 맨 오른편에 위치한 문자의 인덱스 번호를 확인하기 어렵다.
그래서 파이썬에서는 -1을 마지막 문자의 인덱스로 사용한다. 
즉, 맨 오른편의 인덱스는 -1이고, 왼쪽으로 이동하면서 -2, -3 등으로 인덱스를 사용한다.

In [10]:
a_food[-1]

'p'

In [11]:
a_food[-2]

'a'

In [12]:
a_food[-3]

'b'

#### 문자열의 길이와 인덱스

문자열의 길이보다 같거나 큰 인덱스를 사용하면 오류가 발생한다.
문자열의 길이는 `len()` 함수를 이용하여 확인할 수 있다.

In [13]:
len(a_food)

5

In [14]:
a_food[5]

IndexError: string index out of range

마지막 문자열의 인덱스는 -1 또는 문자열의 길이에서 1을 뺀값이다.

In [ ]:
a_food[len(a_food)-1]

In [15]:
a_food[-1]

'p'

### 슬라이싱

문자열에서 하나의 문자가 아닌 특정 구간에 위치한 문자열을 함께 
추출하고자 할 경우 슬라이싱을 사용한다. 
슬라이싱은 다음과 같이 실행한다.

```python
문자열변수[시작 인덱스 : 끝 인덱스 : 스텝(step)]
```

* 시작 인덱스: 추출 구간 시작 인덱스
* 끝 인덱스: 추출 구간 끝 인덱스에 하나 더한 값
* 스텝: 지정된 구간에서 몇 개씩 건너 뛰며 추출할 것인지 지정
    - 스텝은 생략될 수 있으며, 그런 경우 기본값으로 1을 사용.
        
        ```python
        문자열변수[시작 인덱스 : 끝 인덱스]
        ```

In [16]:
a_food

'kebap'

`kebap`에서 `ke` 부분을 추출하고 싶다면 다음과 같이 하면 된다:

In [17]:
a_food[0 : 2 : 1]

'ke'

즉, 문자열 처음부터 2번 인덱스 전까지, 즉 두 번째 문자까지 모두 추출하는 것이다. 
스텝이 1이면 스텝 부분을 없애도 된다.

In [18]:
a_food[0 : 2]

'ke'

하나씩 건너서 추출하려면 다음처럼 하면 된다:

In [19]:
a_food[0 : -1 : 2]

'kb'

시작 인덱스, 끝 인덱스, 스텝 각각의 인자가 경우에 따라 생략될 수도 있다. 
그럴 때는 각각의 위치에 기본값(default)이 들어 있는 것으로 처리되며, 각 자리의 기본값은 다음과 같다.

* `시작 인덱스`의 기본값 = `0`
* `끝 인덱스`의 기본값 = 문자열의 길이
* `스텝`의 기본값 = `1`

In [20]:
a_food[0 : 2]

'ke'

In [21]:
a_food[: 2]

'ke'

In [22]:
a_food[: 4 : 2]

'kb'

In [23]:
a_food[ : : 2]

'kbp'

양수와 음수를 인덱스로 섞어서 사용할 수도 있다.

In [24]:
a_food[ : -1 : 2]

'kb'

**주의**: -1은 문자열의 끝 인덱스를 의미한다.

끝 인덱스가 문자열의 길이보다 클 수도 있다. 
다만 문자열의 길이 만큼만 문자를 확인한다. 

In [25]:
a_food[: 10]

'kebap'

아래와 같이 아무 것도 입력하지 않으면 해당 문자열 전체를 추출한다.

In [26]:
a_food[:]

'kebap'

시작 인덱스 값이 끝 인덱스 값보다 같거나 작아야 제대로 추출한다. 
그렇지 않으면 공문자열이 추출된다.

In [27]:
a_food[3 : 1]

''

이유는 슬라이싱은 기본적으로 작은 인덱스에 큰 인덱스 방향으로 확인하기 때문이다.
역순으로 추출하고자 한다면 스텝을 음수로 사용하면 된다.

In [28]:
a_food[3 : 1 : -1]

'ab'

In [29]:
a_food[-1 : : -1]

'pabek'

### 문자열 메서드

인덱스와 슬라이싱의 기능을 이해하였다면 이제 `text` 변수에 할당된 문자열에서 `">$"`라는
문자열의 시작위치를 알아내기만 하면 된다. 
아주 간단한 방법이 있다. 0번부터 시작해서 주욱 세어가면서 `">$"`의 시작 문자인
`">"`의 인덱스를 확인하면 된다. 
하지만 이런 방식은 아래와 같은 이유로 매우 위험하다.

* 셈이 틀릴 수 있다.
* 문자열이 길 경우 셈 자체가 불가능할 수 있다.
* 문자열이 조금만 변경되어도 새로 처음부터 세어야 하기 때문에 경우에 따라 재활용이 불가능하다.

이런 문제를 해결하는 좋은 방법이 있다. 
바로 `find()`라는 문자열 메서드를 활용하면 된다.

#### `find()` 메서드 활용하기

In [30]:
text.find(">$")

232

이제, 찾고자 하는 `">$"` 문자열이 232번 인덱스에서 시작한다는 것을 알았다.
따라서 커피콩의 가격정보는 인덱스가 2보다 큰 234번이고 거기서부터 길이가 4인
부분문자열에 담겨 있게 된다.

In [31]:
print(text[234: 238])

5.01


하지만, 여기서 234를 사용하기 보다는 `find()` 메서드를 
직접 활용하는 것이 더욱 좋다.

In [32]:
price_index = text.find(">$") + 2

In [33]:
bean_price_str = text[price_index : price_index + 4]

In [34]:
print(bean_price_str)

5.01


`find()` 메서드처럼 문자열 자료형에만 사용하는 함수들이 있다. 
이와같이 특정 자료형에만 사용하는 함수들을 __메서드__(method)라 하며,
특정 자료형을 잘 다루기 위해서는 
어떤 경우에 어떤 메서드를 유용하게 활용할 수 있는지를 잘 파악해두는 것이 매우 중요하다.

#### 메서드 호출 방법

앞서 `find()` 메서드를 호출하는 방법을 기억해야 한다. 
메서드는 일반적인 함수들과는 달리, 특정 자료형의 값이 먼저 언급된 다음에
호출된다. 

```python
text.find("<$")
```

**주의:** 메서드의 호출방식은 다른 자료형의 경우에도 동일하다. 

`find()` 메서드 이외에 문자열과 관련된 메서드는 매우 많다. 
여기서는 가장 많이 사용되는 메서드 몇 개를 소개하고자 한다.

* `strip()`
* `split()`
* `replace()`
* `upper()`
* `lower()`
* `capitalize()`
* `title()`
* `startswith()`
* `endswith()`

예제를 통해 각 메서드의 활용법을 간략하게 확인한다.
먼저 `week_days` 변수에 요일들을 저장한다.

In [35]:
week_days = " Mon, Tue, Wed, Thu, Fri, Sat, Sun "

#### `strip()` 메서드

문자열의 양 끝을 지정한 문자열 기준으로 삭제하는 방식으로 정리한다.
예를 들어, 문자열 양끝에 있는 스페이스를 삭제하고자 할 경우 아래와 같이 실행한다.

In [36]:
week_days.strip(" ")

'Mon, Tue, Wed, Thu, Fri, Sat, Sun'

`strip()` 메서드를 인자 없이 호출하는 경우와 동일하다.

In [37]:
week_days.strip()

'Mon, Tue, Wed, Thu, Fri, Sat, Sun'

#### `split()` 메서드

지정된 부분문자열을 기준으로 문자열을 쪼개어 문자열들의 리스트로 반환한다.
리스트 자료형은 이후에 자세히 다룬다. 여기서는 기본적으로 알고 있는 내용으로 이해하면 된다.
아래 예제는 `", "`, 즉 콤마와 스페이스를 기준으로 문자열을 쪼갠다.

In [38]:
week_days.split(", ")

[' Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun ']

두 개 이상의 메서드를 조합해서 활용할 수도 있다.
예를 들어, `strip()` 메서드를 먼저 실행한 다음에 그 결과에 `split()` 메서드를 실행하면
좀 더 산뜻한 결과를 얻을 수 있다.

In [39]:
week_days.strip(" ").split(", ")

['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']

#### `replace()` 메서드

하나의 문자열을 다른 문자열로 대체한다.
예를 들어, `" Mon"`을 `"Mon"`으로 대체할 경우 아래와 같이 실행한다.

In [40]:
week_days.replace(" Mon", "Mon")

'Mon, Tue, Wed, Thu, Fri, Sat, Sun '

#### `upper()` 메서드

모든 문자를 대문자로 변환시킨다.

In [41]:
week_days.upper()

' MON, TUE, WED, THU, FRI, SAT, SUN '

In [42]:
week_days.strip().upper()

'MON, TUE, WED, THU, FRI, SAT, SUN'

#### `lower()` 메서드

모든 문자를 소문자로 변환시킨다.

In [43]:
week_days.lower()

' mon, tue, wed, thu, fri, sat, sun '

In [44]:
week_days.strip().lower()

'mon, tue, wed, thu, fri, sat, sun'

In [45]:
week_days.strip().lower().split(", ")

['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']

#### `capitalize()` 메서드

문자열의 제일 첫 문자를 대문자로 변환시킨다.
아래 예제는 변화가 없어 보인다. 이유는 첫 문자가 스페이스이기 때문이다.

In [46]:
week_days.capitalize()

' mon, tue, wed, thu, fri, sat, sun '

공백을 없애면 첫문자가 대문자고 바뀐다.

In [47]:
week_days.strip().capitalize()

'Mon, tue, wed, thu, fri, sat, sun'

#### `title()` 메서드

각각의 단어의 첫 문자를 대문자로 변환시킨다.
영어로 쓰여진 신문 기사, 책 제목의 타이틀에서 각 단어의 첫 알파벳이 대문자로 쓰여지는 것과 동일하게 작동한다.

In [48]:
week_days.title()

' Mon, Tue, Wed, Thu, Fri, Sat, Sun '

In [49]:
week_days.strip().title()

'Mon, Tue, Wed, Thu, Fri, Sat, Sun'

#### `startswith()` 메서드

문자열이 특정 문자열로 시작하는지 여부를 판단해준다.

In [50]:
week_days.startswith(" M")

True

#### `endswith()` 메서드

문자열이 특정 문자열로 끝나는지 여부를 판단해준다.

In [51]:
week_days.endswith("n ")

True

#### `count()` 메서드

문자열 안에 특정 문자열이 몇 번 출현하는가를 세어준다.
예를 들어, 문자열이 몇 줄로 이루어졌는가를 
확인하려면 `count()` 메서드를 이용하여 
줄바꿈 기호(`\n`)가 사용된 횟수를 세면 된다.

In [52]:
text.count('\n')

7

## 변경 불가능한 자료형

파이썬의 문자열 자료형의 값들은 __변경 불가능__(immutable)이다.
앞서  `week_days`에 할당된 문자열에 다양한 메서드를 적용하여 새로운 문자열을 생성하였지만
`week_days`에 할당된 문자열 자체는 전혀 변하지 않았음을 아래와 같이 확인할 수 있다.

In [53]:
week_days

' Mon, Tue, Wed, Thu, Fri, Sat, Sun '

즉, 앞서 소개한 문자열 메서드는 기존 문자열을 변경하는 것이 아니라 기존 문자열을 이용하여 항상 새로운 
문자열을 생성한다. 
이와 같이 한 번 정해지면 절대 변경이 불가능한 자료형을 변경 불가능(immutable) 자료형이라 부른다.

주어진 문자열을 이용하여 새로운 문자열을 생성하고 활용하려면 새로운 변수에 저장하여 활용해야 한다.

In [54]:
stripped_week_days = week_days.strip()

In [55]:
stripped_week_days

'Mon, Tue, Wed, Thu, Fri, Sat, Sun'

In [56]:
week_days_1 = week_days.replace(' Mon', 'Mon')
week_days_1

'Mon, Tue, Wed, Thu, Fri, Sat, Sun '

### `str()` 함수

`str()` 함수를 이용하여 다양한 값을 문자열로 변환할 수 있다.

__주의사항:__ `str`은 문자열 자료형을, `str()`은 문자열을 반환하는 함수를 가리킨다.

In [57]:
a = 5.6
s = str(a)

type(s)

str

### 역슬래시(&bsol;) 기능

__주의사항:__ 윈도우 브라우저에서는 역슬래시(&bsol;)가 원화 통화기호(&#8361;) 모양으로 보이지만,
동일하게 기능한다.

역슬래시 문자(&bsol;)는 특수한 기능을 수행한다.
예를 들어, 줄바꿈을 의미하는 문자 `\n`, 
탭을 의미하는 문자 `\t` 등에서 역슬래시가 특수한 기능을 수행한다. 
따라서 역슬래시 자체를 문자열에 포함할 때 조심해야 한다.
예를 들어, 아래와 같은 문자열을 사용하고자 한다고 가정하자.

```python
"12\34"
```

그런데 그냥 아래와 같이 지정하면 다르게 작동한다.
이유는 `\3`이 특수한 문자를 표현하도록 역슬래시가 기능하기 때문이다.

In [58]:
s = '12\34'

print(s)

12


In [59]:
s = '\3'

print(s)




따라서 `12\34`로 출력되게 하려면 역슬래시의 기능을 해제해야 하며,
그러기 위해 역슬래시를 두 번 적어주면 된다.
그러면 첫째 역슬래시 뒤에 나오는 역슬래시의 기능을 해제해서 지정한 문자열로 처리된다.

__참고:__ 이처럼 특정 기능을 작동하지 못하도록 하는 것을 영어로 __이스케이프__(escape)라고 부른다.

In [60]:
s = '12\\34'

print(s)

12\34


그런데 문자열 안에 많은 역슬래시가 포함되어 있다면 
이런 방식은 매우 불편하다. 
하지만 문자열 앞에 영어 알파벳 `r`을 추가하면 간단하게 해결된다.

In [61]:
s = r'this\has\no\special\characters'

print(s)

this\has\no\special\characters


### 문자열 연산: 덧셈과 정수 곱셈

두 문자열을 더하면 두 문자열을 이어서 붙인다.

In [62]:
a = 'this is the first half '
b = 'and this is the second half'

a + b

'this is the first half and this is the second half'

문자열과 정수를 곱하면 해당 정수만큼 복사되어 길어진다.

In [63]:
a * 2

'this is the first half this is the first half '

### 문자열 템플릿

__문자열 템플릿__은 문자열 안에 일부 변하는 값을 지정할 수 있도록 선언된 문자열이다.
예를 들어, 아래 템플릿은 세 개의 값을 임의로 지정할 수 있도록 준비되어 있다.

In [64]:
template = '{0:.2f} {1:s}는 {2:d} 미국달러에 해당한다.'
template

'{0:.2f} {1:s}는 {2:d} 미국달러에 해당한다.'

__참고:__ 중괄호 안에 사용된 숫자와 기호의 의미는 다음과 같다.

* `0:.2f` - `format()` 메서드의 첫째 인자인 부동소수점이 자리하며 소수점 이하 두 자리까지 표기
* `1:s` - `format()` 메서드의 둘째 인자인 문자열이 자리하는 자리
* `2:d` - `format()` 메서드의 셋째 인자인 정수가 위치하는 자리

문자열 템플릿에 지정된 수 만큼의 값을 `format()` 메서드를 이용하여 입력하여
새로운 문자열을 생성할 수 있다.
단, `format()` 메서드에 사용되는 인자의 순서는 지정된 순서대로 정해져야 한다.

예를 들어, `template` 변수가 가리키는 문자열 템플릿의 세 위치에 차례대로
부동소수점, 문자열, 정수를 입력해야 하며, 아래와 같이 
차례대로 인자로 사용하면 된다.

In [65]:
template.format(4.5560, '아르헨티나 페소', 1)

'4.56 아르헨티나 페소는 1 미국달러에 해당한다.'

__참고:__
콜론(`:`)을 포함하여 그 이후에 해당하는 부분은 입력될 값들의 자료형을 안내하는 역할을 수행한다.
하지만 의무사항은 아니며 다만, 사용되는 값에 대한 정보를 제공하거나
아니면 보다 서식을 갖춘 문자열이 출력하도록 사용된다.

In [66]:
template = '{0} {1}는 {2} 미국달러에 해당한다.'

template.format(4.5560, '아르헨티나 페소', 1)

'4.556 아르헨티나 페소는 1 미국달러에 해당한다.'

#### f-문자열

`format()` 메서드를 사용하는 대신에 
요즘에는 f-문자열을 보다 많이 사용한다.
이유는 보다 편리한 사용성 때문이다.
f-문자열은 문자열 앞에 영어 알파벳 f를 추가하기만 하면 되며 
문자열 안에 변수를 중괄호로 감싸 직접 대입한다. 

In [67]:
a = 4.5560
b = '아르헨티나 페소'
c = 1

In [68]:
template = f'{a:.2f} {b:s}는 {c:d} 미국달러에 해당한다.'

template

'4.56 아르헨티나 페소는 1 미국달러에 해당한다.'

여기서도 콜론 부분은 생략해도 된다.

In [69]:
template = f'{a} {b}는 {c} 미국달러에 해당한다.'

template

'4.556 아르헨티나 페소는 1 미국달러에 해당한다.'

## 연습문제

애완동물의 목록을 할당받는 `pets` 변수가 아래와 같이 선언되어 있다. 

In [70]:
pets = 'dog cat hedgehog pig swan fish bird'

### 연습

애완동물의 종류를 의미하는 단어의 첫알파벳을 대문자로 바꾸려면 어떻게 해야 하는가? 
단, 특정 메서드를 사용하여 한 줄 코드로 작성해야 한다.

견본답안:

In [71]:
pets.title()

'Dog Cat Hedgehog Pig Swan Fish Bird'

### 연습

`pets`으로부터 대문자 `C` 문자 하나를 추출하라.

견본답안:

In [72]:
pets.title()[4]

'C'

### 연습

`hedgehog`을 추출하려면?

견본답안:

In [73]:
pets[8 : 16]

'hedgehog'

### 연습 (이전 문제 이어서)

`hdeo`을 추출하려면?

견본답안:

In [74]:
pets[8 : 16 : 2]

'hdeo'

### 연습

`gohegdeh`을 추출하려면?

견본답안:

In [75]:
pets[15: 7 : -1]

'gohegdeh'

### 연습

`dogs`와 `cats` 두 개의 변수가 다음과 같이 선언되었다.

In [76]:
dogs, cats = '8', '4'

* 강아지와 고양이를 몇 마리씩 갖고 있는지 확인하는 방법은?
* 강아지가 고양이보다 몇 마리 더 많은지 확인하는 방법은?

견본답안:

In [77]:
print(int(dogs))

8


In [78]:
print(int(cats))

4


In [79]:
print(abs(int(dogs) - int(cats)))

4
